lib/lookup/src/lookup_v2/mod.rs

```diff
@@ -186,6 +200,33 @@ impl<'a, 'b: 'a> From<&'b OwnedSegment> for BorrowedSegment<'a> {
     }
 }
 
+impl Display for OwnedSegment {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            OwnedSegment::Field(field) => {
+                static UNESCAPED_PATTERN: Lazy<Regex> =
+                    Lazy::new(|| Regex::new(r"^[A-Za-z_0-9@]*$").unwrap());
+                if !UNESCAPED_PATTERN.is_match(field) {
+                    let mut string = String::from('"');
+                    string.reserve(field.as_bytes().len());
+                    for c in field.chars() {
+                        if matches!(c, '"' | '\\') {
+                            string.push('\\');
+                        }
+                        string.push(c);
+                    }
+                    string.push('"');
+                    write!(formatter, ".{}", string)
+                } else {
+                    write!(formatter, ".{}", field)
+                }
+            }
+            OwnedSegment::Index(index) => write!(formatter, "[{}]", index),
+            OwnedSegment::Invalid => write!(formatter, ".<invalid>"),
```

Comment:
I'm not sure why the `Invalid` variant is a thing, and why we don't fail when building a `OwnedSegment` instead.

As long as the variant exists though, I opted to choose the string `.<invalid>` which will again produce a `Invalid` variant when being parsed.